File: bitonic_sort.py
import math
import random

'''
TODO:
Alter input data to match testing parameters
Optimize cache size and table data structure
Alter functions to work for new table structure
Add more testing sims
Add encryption and decryption functions
Change from random module to secrets module
'''

class Server():
    def __init__(self, data):
        self.data = data
        self.access_log = []
    
    
class Client():
    #use server and table size as inputs
    def __init__(self, server):
        #Permutation stores the current mapping of the data
        #Cache stores recently accessed values
        self.permutation = {}
        self.server = server
        self.cache = []
        
        #create initial map
        for i in range(len(server.data)):
            #change server.data to whatever data index used in server
            self.permutation[i] = server.data[i]
    
    #gets a row from the server with relation to the random permutation
    def queryServer(self, index):
        row = self.server.data[self.permutation[index]]
        self.cache.append[row]
        yield row
        if len(self.cache) == 5:
            self.cache = []
            self.obvshuffle()
        
        #compares the permutations' values, if switch, reflect in 
    def compAndSwap(self, data, a, i, j, dire):
        if (dire==1 and a[i] > a[j]) or (dire==0 and a[i] < a[j]):
            a[i],a[j] = a[j],a[i]
            data[i],data[j] = data[j], data[i]
            self.permutation[i],self.permutation[j] = self.permutation[j],self.permutation[i]
    
    def bitonicMerge(self, a, data, low, cnt, dire):
        '''
        Recursively sorts the bitonic sequence
        If dire = 1, sorts ascending, else descending
        low is the starting index, cnt is the # of elements in the section
        '''
        if cnt > 1:
          k = cnt//2
          for i in range(low , low+k):
            self.compAndSwap(a, data, i, i+k, dire)
          self.bitonicMerge(a, data, low, k, dire)
          self.bitonicMerge(a, data, low+k, k, dire)
 
    def bitonicSort(self, a, data, low, cnt, dire):
        '''
        Produces a bitonic sequence by recursively splitting it and sorting it them in opposite sorting orders
        Calls bitonicMerge on each part, sorting them in the determined order
        '''
        if cnt > 1:
            #halves the # of entries for each recursion
            k = cnt//2
            self.bitonicSort(a, data, low, k, 1)
            self.bitonicSort(a, data, low+k, k, 0)
            self.bitonicMerge(a, data, low, cnt, dire)
    

    def obvshuffle(self): 
        data = self.server.data.copy()
        N = len(data)
        
        permutation = [i for i in range(N)]
        random.shuffle(permutation)
        #assert that the input size is a power of 2
        if math.log2(N) % 1 != 0:
            raise ValueError("Input is not a power of 2")
        print(self.permutation)
        print(permutation)
        print(self.server.data)
        self.bitonicSort(permutation, data, 0, N, 1)
        self.server.data = data
        print(self.permutation)
        
#debug
if __name__ == "__main__":
    n = 16
    data = [random.randint(1,100) for i in range(n)]
    server = Server(data)
    client = Client(server)
    client.obvshuffle()

File: bitonic_sort.cpp
#include "bitonic_sort.h"
#include <vector>
#include <string>

//compares the values and replaces the values that are different

//Server Methods--
Server::Server(vector<string> data, int input_size) {
    storage = data;
    size = input_size;
}

string Server :: get_value(int index) {
    return storage[index];
}

int Server :: set_value(int index, string value) {
    storage[index] = value;
    return 0;
}

//Client Methods--
Client::Client(Server* new_server) {
  server = new_server;
}

//without encryption
void Client :: comp_values(int i, int j, int dir) {
    string a = server->get_value(i);
    string b = server->get_value(j);
    if (dir==(a>b)) {
        server->set_value(j, a);
        server->set_value(i, b);
    }
    else {
        server->set_value(i, a);
        server->set_value(j, b);
    }
}
 
//Recursively sorts the bitonic sequence
//If dire = 1, sorts ascending, else descending
//low is the starting index, cnt is the # of elements in the section
void Client :: bitonic_merge(int low, int cnt, int dir) {
    if (cnt>1) {
        int k = cnt/2;
        for (int i=low; i<low+k; i++) {
            comp_values(i, i+k, dir);
	}
        bitonic_merge(low, k, dir);
        bitonic_merge(low+k, k, dir);
    }
}
 
//Produces a bitonic sequence by recursively splitting it and sorting it them in opposite sorting orders
//Calls bitonic_merge on each part, sorting them in the determined order
//Low is the starting index, count is the size of each split, dir is the direction of the sort
void Client :: bitonic_sort(int low, int cnt, int dir) {
    if (cnt>1) {
        int k = cnt/2;
        bitonic_sort(low, k, 1);
        bitonic_sort(low+k, k, 0);
        bitonic_merge(low, cnt, dir);
    }
}
 
//calls bitonic sort in ascending order with Length N
//up = 1 is ascending, 0 is descending
void Client :: sort(int N, int up) {
    bitonic_sort(0, N, up);
}

string Client :: encrypt(string data) {
  string result = "";
  int l = data.length();
  for (int i = 0; i < l; i++) {
    result += data[i] ^ encryption_key[i];
  }
  return result;
}

string Client :: decrypt(string data) {
  string result = "";
  int l = data.length();
  for (int i = 0; i < l; i++) {
    result += data[i] ^ encryption_key[i];
  }
  return result;
}




